components/web/services/org.wso2.carbon.apimgt.authenticator/src/main/java/org/wso2/carbon/apimgt/authenticator/utils/AuthUtil.java

```diff
@@ -128,20 +135,61 @@ public static ErrorDTO getErrorDTO(ErrorHandler errorHandler, HashMap<String, St
     }
 
     /**
-     * @param cookie Cookies  header which contains the access token
-     * @return partial access token present in the cookie.
+     * Method used to extract refresh token from headers.
+     * @param headers  headers which contains the access token
+     * @return refresh token present in the cookie and authorization header..
      */
-    public static String extractPartialAccessTokenFromCookie(String cookie) {
-        cookie = cookie.trim();
-        String[] cookies = cookie.split(";");
-        String token = Arrays.stream(cookies).filter(name -> name.contains(AuthenticatorConstants.REFRESH_TOKEN))
-                .findFirst().get();
-        if (token.split("=").length == 2) {
-            return token.split("=")[1];
+    public static String extractRefreshTokenFromHeaders(Headers headers) {
+        String authHeader = headers.get(AuthenticatorConstants.AUTHORIZATION_HTTP_HEADER);
+        String refreshToken = "";
+        if (authHeader != null) {
+            authHeader = authHeader.trim();
+            if (authHeader.toLowerCase(Locale.US).startsWith(AuthenticatorConstants.BEARER_PREFIX)) {
+                // Split the auth header to get the access token.
+                String[] authHeaderParts = authHeader.split(" ");
+                if (authHeaderParts.length == 2) {
+                    refreshToken = authHeaderParts[1];
+                } else if (authHeaderParts.length < 2) {
+                    return null;
+                }
+            }
         }
-        return null;
-    }
+        String cookie = headers.get(AuthenticatorConstants.COOKIE_HEADER);
+        if (cookie != null) {
+            cookie = cookie.trim();
+            String[] cookies = cookie.split(";");
+            String token = Arrays.stream(cookies).filter(name -> name.contains(AuthenticatorConstants.REFRESH_TOKEN_2))
+                    .findFirst().orElse("");
+            if (token.split("=").length == 2) {
+                refreshToken += token.split("=")[1];
```

Comment:
instead of doing token split two times how about assign it in first place and use it